### PR TITLE
Make unix signal-handler signal-safe

### PR DESCRIPTION
**Pull request by hulthe**
Other changes:
- Re-enable the signal handler in release-builds.
- Disable backtrace printing by default since it's not signal-safe.
- Add `MULLVAD_BACKTRACE_ON_FAULT` env variable to enable backtracing (Yes, that makes it not signal safe! But something we are willing to have as opt-in due to how it can help us debug stuff).
- Remove the alternate signal stack. The reasons for this are:
  - Setting up an alt-stack in a safe way is not trivial, our previous
    attempt was unsound in the presence of stack overflows. It can be
    done safely with mmap, but would require careful review.
  - The alt-stack is thread-local, meaning it would need to be
    initialized on a per-thread basis. We would need to hook into tokio
    and std::thread::spawn to be able to get good coverage, and even
    then there would no good way to ensure that *all* threads have an
    alternate stack, except that...
  - Rust (by default) allocates an alternate stack for every thread.
    Unfortunately, the prescence of Go code in our linked binary
    disables this feature. IMO, we should strive towards not having any
    Go code linked into the daemon for this reason.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)

### CODE REVIEW

**Review by hulthe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: 0 of 4 files reviewed, 1 unresolved discussion

___
*[`mullvad-daemon/src/logging.rs` line 121 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OB_YKWB8aaX7WzCOj0u:-OB_YKWB8aaX7WzCOj0v:bb7pa6j) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/f18c5e16c79371e4cfdc504dd0362f0e4c4f5b29/mullvad-daemon/src/logging.rs#L121)):*
> ```XML
>         #[cfg(unix)]
>         crate::exception_logging::set_log_file(
>             std::ffi::CString::new(log_file.as_os_str().as_encoded_bytes()).unwrap(),
> ```

todo

**Review by hulthe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: 0 of 4 files reviewed, 2 unresolved discussions

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 113 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OB_Y_CwDzO7D1u4VwZD:-OB_Y_CwDzO7D1u4VwZE:ba3sxwj) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/f18c5e16c79371e4cfdc504dd0362f0e4c4f5b29/mullvad-daemon/src/exception_logging/unix.rs#L113)):*
> ```XML
> 
>     let code: c_int = match log_fault_to_file_and_stdout(signum) {
>         // Signal numbers are positive integers
> ```

todo: not sure if this is true? The exit code in practice appears to be e.g. `-11` for SIGSEGV.

Also: calling _exit() with a negative code will return `256 - code`, which is strange to say the least.

**Review by MarkusPettersson98 — COMMENTED** [review]
Very nice work :yellow_heart: 

> Reviewed 3 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-:-OB_noRi5Tj9IT1FsNd2:bgykzai)* status: 1 of 4 files reviewed, 3 unresolved discussions (waiting on @hulthe)

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 177 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OB_kHN0ECjz6X6LdKAL:-OB_kHN0ECjz6X6LdKAM:bz21db4) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/f18c5e16c79371e4cfdc504dd0362f0e4c4f5b29/mullvad-daemon/src/exception_logging/unix.rs#L177)):*
> ```XML
>         Err(_) => {
>             // SIGNAL-SAFETY: formatting an i32 is signal-safe.
>             writeln!(w, "Signal handler triggered by unknown signal: {signum}")?;
> ```

Could you convince me since I do not carry the same intuition as you; is formatting an i32 signal-safe because it fits into the size of a pointer? Or is it because it is allocated on the stack? etc



_Code quote:_
```XML
            // SIGNAL-SAFETY: formatting an i32 is signal-safe.
            writeln!(w, "Signal handler triggered by unknown signal: {signum}")?;
```

**Review by hulthe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: 1 of 4 files reviewed, 3 unresolved discussions (waiting on @MarkusPettersson98)

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 177 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OB_kHN0ECjz6X6LdKAL:-OB_zU5y2QF30SVozRZa:b-jkphrb) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/f18c5e16c79371e4cfdc504dd0362f0e4c4f5b29/mullvad-daemon/src/exception_logging/unix.rs#L177)):*
<details><summary><i>Previously, MarkusPettersson98 (Markus Pettersson) wrote…</i></summary><blockquote>

Could you convince me since I do not carry the same intuition as you; is formatting an i32 signal-safe because it fits into the size of a pointer? Or is it because it is allocated on the stack? etc
</blockquote></details>

That's a very good question, and not easy to answer.

There's no concept in Rust of signal-safety, so the only way to be sure is to look at the implementation. In this case, the implementation of [Display for i32](https://doc.rust-lang.org/src/core/fmt/num.rs.html#210). An implementation is, of course, subject to change.

Formally: A function is signal-safe if it does not interact with any global shared mutable state (e.g. an allocator). Less formally: One *could* intuit (without looking at the source code) that property *should* hold for i32::fmt, since it would be pretty crazy for such a function to allocate, or to have any other global state, but there's no guarantees of that.

In summary: I looked at the source of i32::fmt, and it looks safe to me. That might change in the future, but I doubt it will. The alternative is to not print the signal number at all, or to format it ourselves.

**Review by MarkusPettersson98 — COMMENTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on @hulthe)

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 177 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OB_kHN0ECjz6X6LdKAL:-OBa6qaeBKDK6wpcNtk_:bvtq4c9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/f18c5e16c79371e4cfdc504dd0362f0e4c4f5b29/mullvad-daemon/src/exception_logging/unix.rs#L177)):*
<details><summary><i>Previously, hulthe (Joakim Hulthe) wrote…</i></summary><blockquote>

That's a very good question, and not easy to answer.

There's no concept in Rust of signal-safety, so the only way to be sure is to look at the implementation. In this case, the implementation of [Display for i32](https://doc.rust-lang.org/src/core/fmt/num.rs.html#210). An implementation is, of course, subject to change.

Formally: A function is signal-safe if it does not interact with any global shared mutable state (e.g. an allocator). Less formally: One *could* intuit (without looking at the source code) that property *should* hold for i32::fmt, since it would be pretty crazy for such a function to allocate, or to have any other global state, but there's no guarantees of that.

In summary: I looked at the source of i32::fmt, and it looks safe to me. That might change in the future, but I doubt it will. The alternative is to not print the signal number at all, or to format it ourselves.
</blockquote></details>

Great answer, thanks :+1:

**Review by hulthe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @MarkusPettersson98)

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 113 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OB_Y_CwDzO7D1u4VwZD:-OBaBBqg9yj1iRTlWflJ:b-wq9j4a) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/f18c5e16c79371e4cfdc504dd0362f0e4c4f5b29/mullvad-daemon/src/exception_logging/unix.rs#L113)):*
<details><summary><i>Previously, hulthe (Joakim Hulthe) wrote…</i></summary><blockquote>

todo: not sure if this is true? The exit code in practice appears to be e.g. `-11` for SIGSEGV.

Also: calling _exit() with a negative code will return `256 - code`, which is strange to say the least.
</blockquote></details>

Made the signum-codes positive instead.

**Review by MarkusPettersson98 — DISMISSED** [review]
> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 113 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OB_Y_CwDzO7D1u4VwZD:-OBdYzWm-byBhMqNwFlV:b-dxnmqx) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/f18c5e16c79371e4cfdc504dd0362f0e4c4f5b29/mullvad-daemon/src/exception_logging/unix.rs#L113)):*
<details><summary><i>Previously, hulthe (Joakim Hulthe) wrote…</i></summary><blockquote>

Made the signum-codes positive instead.
</blockquote></details>

:smile: :ok_hand:

**Review by pinkisemils — DISMISSED** [review]
:lgtm:

> Reviewed 1 of 4 files at r1, 1 of 2 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-:-OBkv11f3Jgo9opvRK7F:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

**Review by faern — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @hulthe, @MarkusPettersson98, and @pinkisemils)

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 2 at r5](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OC4Uwu60TAWdqQeGc2q:-OC4Uwu60TAWdqQeGc2r:bk7tcn6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d80f9f851caac261e33a7df2f0cd3d807f5bcd6a/mullvad-daemon/src/exception_logging/unix.rs#L2)):*
> ```XML
> //! Install signal handlers to catch critical program faults and log them. See [`enable`].
> #![warn(clippy::undocumented_unsafe_blocks)]
> ```

This can probably be removed? I fail to make it generate any warnings when I remove it, so I suspect it's a leftover from development? Either way, we should try to avoid suppressing linting as much as possible.

**Review by MarkusPettersson98 — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: all files reviewed, 1 unresolved discussion (waiting on @hulthe)

**Review by hulthe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: all files reviewed, 1 unresolved discussion (waiting on @faern)

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 2 at r5](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OC4Uwu60TAWdqQeGc2q:-OC8UEp18ftQb9U7W54C:bi9dc1i) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d80f9f851caac261e33a7df2f0cd3d807f5bcd6a/mullvad-daemon/src/exception_logging/unix.rs#L2)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

This can probably be removed? I fail to make it generate any warnings when I remove it, so I suspect it's a leftover from development? Either way, we should try to avoid suppressing linting as much as possible.
</blockquote></details>

undocumented_unsafe_blocks is actually disabled by default, this _enables_ that lint.

Related issue: [DES-1459](https://linear.app/mullvad/issue/DES-1459)

**Review by faern — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 2 at r5](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OC4Uwu60TAWdqQeGc2q:-OC922gE8zB26fTxj2n0:byrhr0o) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d80f9f851caac261e33a7df2f0cd3d807f5bcd6a/mullvad-daemon/src/exception_logging/unix.rs#L2)):*
<details><summary><i>Previously, hulthe (Joakim Hulthe) wrote…</i></summary><blockquote>

undocumented_unsafe_blocks is actually disabled by default, this _enables_ that lint.

Related issue: [DES-1459](https://linear.app/mullvad/issue/DES-1459)
</blockquote></details>

Aaah. First time someone fine grained adds more stricter linting maybe! So I just assumed you made it weaker :joy: We should ideally have this be warn/deny in the entire workspace. But that's a large body of work that is completely unrelated to this PR :) I endorse stricter linting so please go ahead with this

**Review by pinkisemils — APPROVED** [review]
> Reviewed 1 of 2 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

**Review by faern — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: 2 of 5 files reviewed, 1 unresolved discussion (waiting on @hulthe, @MarkusPettersson98, and @pinkisemils)

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 64 at r8](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OCIEzaw2OZBL3O7js-9:-OCIEzaw2OZBL3O7js-A:bvm88ke) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcb3ac7fc26f21f6721460042e73b617b8d3ba61/mullvad-daemon/src/exception_logging/unix.rs#L64)):*
> ```XML
> 
>     INIT_ONCE.call_once(|| {
>         if let Ok(override_backtrace) = env::var(ENABLE_BACKTRACE_VAR).map(|v| v == "1") {
> ```

All other boolean environment variables that control the daemon has the enable logic ` != "0"`. Can we keep this so it's consistent? There should only be two possible values, but I'd like all boolean control mechanisms to react the same to setting the value to `"sausage"` if we can.

**Review by hulthe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: 2 of 5 files reviewed, 1 unresolved discussion (waiting on @faern, @MarkusPettersson98, and @pinkisemils)

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 64 at r8](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OCIEzaw2OZBL3O7js-9:-OCIKiao6n9_KROJ307T:bo6vx3s) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcb3ac7fc26f21f6721460042e73b617b8d3ba61/mullvad-daemon/src/exception_logging/unix.rs#L64)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

All other boolean environment variables that control the daemon has the enable logic ` != "0"`. Can we keep this so it's consistent? There should only be two possible values, but I'd like all boolean control mechanisms to react the same to setting the value to `"sausage"` if we can.
</blockquote></details>

My reasoning is that potentially harmful options should fail safe, and default to false if the input is ambiguous. It shouldn't (imo) be possible that `HARMFUL_OPTION=false` actually _enables_ the harmful option. I'll change it if you disagree though.

Our documentation only defines our env variables in terms of "1" and "0" anyway, so there's no deviation from spec.

**Review by MarkusPettersson98 — APPROVED** [review]
> Reviewed 1 of 1 files at r7, 2 of 3 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: all files reviewed, 1 unresolved discussion (waiting on @faern)

**Review by hulthe — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: 3 of 5 files reviewed, 1 unresolved discussion (waiting on @faern and @MarkusPettersson98)

___
*[`mullvad-daemon/src/exception_logging/unix.rs` line 64 at r8](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172#-OCIEzaw2OZBL3O7js-9:-OCbt9BG2SNy3IXFW2iP:bh7848x) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/fcb3ac7fc26f21f6721460042e73b617b8d3ba61/mullvad-daemon/src/exception_logging/unix.rs#L64)):*
<details><summary><i>Previously, hulthe (Joakim Hulthe) wrote…</i></summary><blockquote>

My reasoning is that potentially harmful options should fail safe, and default to false if the input is ambiguous. It shouldn't (imo) be possible that `HARMFUL_OPTION=false` actually _enables_ the harmful option. I'll change it if you disagree though.

Our documentation only defines our env variables in terms of "1" and "0" anyway, so there's no deviation from spec.
</blockquote></details>

Went ahead and changed it back

**Review by faern — APPROVED** [review]
> Reviewed 2 of 3 files at r12.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/7172)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved
